pkg/ext-proc/backend/pod_controller.go

```diff
@@ -0,0 +1,197 @@
+	}
+
+	logger.Info("Started workers")
+	<-ctx.Done()
+	logger.Info("Shutting down workers")
+
+	return nil
+}
+
+// runWorker is a long-running function that will continually call the
+// processNextWorkItem function in order to read and process a message on the
+// workqueue.
+func (c *PodController) runWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+// processNextWorkItem will read a single work item off the workqueue and
+// attempt to process it, by calling the syncHandler.
+func (c *PodController) processNextWorkItem(ctx context.Context) bool {
+	logger := klog.FromContext(ctx)
+	objRef, shutdown := c.workqueue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	// We call Done at the end of this func so the workqueue knows we have
+	// finished processing this item. We also must remember to call Forget
+	// if we do not want this work item being re-queued. For example, we do
+	// not call Forget if a transient error occurs, instead the item is
+	// put back on the workqueue and attempted again after a back-off
+	// period.
+	defer c.workqueue.Done(objRef)
+
+	// Do work and update the local datastore
+	err := c.updateDatastore(objRef)
+	if err == nil {
+		// If no error occurs then we Forget this item so it does not
+		// get queued again until another change happens.
+		c.workqueue.Forget(objRef)
+		logger.Info("Successfully synced", "objectName", objRef)
+		return true
+	}
+
+	// there was a failure so be sure to report it.  This method allows for
+	// pluggable error handling which can be used for things like
+	// cluster-monitoring.
+	utilruntime.HandleErrorWithContext(ctx, err, "Error syncing; requeuing for later retry", "objectReference", objRef)
+	// since we failed, we should requeue the item to work on later.  This
+	// method will add a backoff to avoid hotlooping on particular items
+	// (they're probably still not going to work right away) and overall
+	// controller protection (everything I've done is broken, this controller
+	// needs to calm down or it can starve other useful work) cases.
+	c.workqueue.AddRateLimited(objRef)
+	return true
+}
+
+func (c *PodController) enqueuePods(obj interface{}) {
+	// Cant select on labels without a selector
+	if c.datastore.LLMServerPool == nil {
+		return
+	}
+
+	if object, ok := obj.(metav1.Object); ok {
+		if _, ok := c.datastore.Pods.Load(object.GetName()); ok || c.datastore.LabelsMatch(object.GetLabels()) {
+			if objectRef, err := cache.ObjectToName(obj); err != nil {
+				utilruntime.HandleError(err)
+				return
+			} else {
+				c.workqueue.Add(objectRef)
+			}
+		}
+	}
+}
+
+func (c *PodController) updateDatastore(objName cache.ObjectName) error {
+	pod, err := c.podLister.Pods(objName.Namespace).Get(objName.Name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			//Delete pod from datastore in a threadsafe way
+			return nil
+		}
+		return err
+	}
+
+	if !c.datastore.LabelsMatch(pod.ObjectMeta.Labels) {
+		c.datastore.Pods.Delete(pod.Name)
```

Comment:
we should also delete it from the store if the pod is marked for deletion (deletionTimestamp is set)